File: python/src/advent_utils.py
"""Utility functions for Advent of Code."""

from collections.abc import Sequence
from pathlib import Path
from typing import Protocol, override


class Stringable(Protocol):
    """Defines a type that could be converted to a string."""

    @override
    def __str__(self) -> str: ...


def read_input(path: str) -> list[str]:
    """Read input file into a polars DataFrame."""
    with Path(path).open() as f:
        return [s.strip() for s in f]


class AdventSolution(Protocol):
    """Represents a result of an Advent of Code problem."""

    @staticmethod
    def part_1(lines: Sequence[str]) -> Stringable:
        """Return the solution for day 1."""
        ...

    @staticmethod
    def part_2(lines: Sequence[str]) -> Stringable:
        """Return the solution for day 2."""
        ...


def run(result: type[AdventSolution], path: str) -> None:
    """Run the solution for a given day."""
    print(f"From {path} =>")  # noqa: T201
    inp = read_input(path)
    day1 = result.part_1(inp)
    print(f"\tPart 1: {day1}")  # noqa: T201
    try:
        day2 = result.part_2(inp)
        print(f"\tPart 2: {day2}")  # noqa: T201
    except AttributeError:
        print("\tPart 2: NOT IMPLEMENTED")  # noqa: T201

File: ocaml/bin/advent_utils.ml
let read_file path =
  let chan = open_in path in
  let rec append_lines acc =
    try
      let line = input_line chan in
        append_lines (line :: acc)
    with End_of_file ->
      close_in chan;
      List.rev acc
    in
    append_lines []

type adventSolution = {
  part_1: string list -> int;
  part_2: string list -> int;
}

let run path solution =
  let lines = read_file path in
    lines |> solution.part_1 |> string_of_int |> String.cat "\tPart 1 => " |> print_string;
    lines |> solution.part_2 |> string_of_int |> String.cat "\tPart 2 => " |> print_string

